Live Weather and Styling

GammaGames opened the pull request:
Added the option to get live weather using weather underground, from #5 
Added alternate icons that change between play and pause to indicate #6 (not in options, just happens by default, could be added)
CSS to make the range, radio, and checkbox inpuits look nicer, renamed the css file to reflect this
town tune page has colors
Volume slider now goes in increments of 0.05 instead of 0.1

kowsen commented:
I'm not a git expert, but I think if you rebase this pull request based on the current state of the JdotCarver repo, it'll make the changes easier to see and should fix the conflicts.

kowsen commented on scripts/AudioManager.js:
```diff
@@ -0,0 +1,112 @@
+
+	addEventListener("hourMusic", playHourlyMusic);
+
+	addEventListener("weatherMusic", playWeatherMusic);
```
I feel like it would be better to handle determining what weather's music we want to play in StateManager rather than AudioManager. For example, if it is snowing and we're on the live weather option, just send "new-leaf-snowing" as the game in a normal hourMusic message. This setup clutters the system with two new events, and means if we want to change the way music plays, we would need to do it in multiple places - and I don't see any advantages to determining what music should be played in AudioManager that offset this.

GammaGames replied:
This is a very good point, and it would work a lot nicer if there were the WeatherManager object like you referenced later. To get the weather working I was modifying it like I had before, this is a much better solution

kowsen commented on scripts/BadgeManager.js:
```diff
@@ -0,0 +1,54 @@
+	chrome.browserAction.setBadgeBackgroundColor({ color: [57, 230, 0, 255] });
+}
+
+function clearBadge() {
```
These functions at the bottom should not be in the global namespace. If the listeners for the events are not managing the state of the badge correctly, there's some issue with the logic firing those events.

kowsen commented on scripts/AudioManager.js:
```diff
@@ -0,0 +1,112 @@
+		}
+		else {
+			audio.pause();
+			clearBadge();
```
Changing the state of the chrome badge should not be handled through here - BadgeManager should pick up the "hourMusic" or "pause" event on its own - the Audio view definitely should NOT be controlling the Badge view.

kowsen commented on scripts/StateManager.js:
```diff
@@ -0,0 +1,168 @@
+			}
+			else if(isLive()) {
+				if(options.music == 'new-leaf-live') {
+					updateWeatherCond(options.zipCode, options.countryCode, function(response) {
```
It would be cleaner to make WeatherManager its own object with a callback for when the weather changes, like TimeKeeper is. It makes things easier to read and debug if the method to retrieve the weather is abstracted away from the main logic of StateManager.

kowsen commented on scripts/Utility.js:
```diff
@@ -0,0 +1,56 @@
+}
+
+// get current weather conditions using openweathermap: http://openweathermap.org/current
+function updateWeatherCond(zip, country, cb) {
```
I don't think updateWeatherCond or getSrc should be Utility functions. My vision for this file was to be functions that are useful to multiple parts of the program, and thus belong in the global namespace. These two should only be seen by individual parts of the program, and should probably be encapsulated in the objects that use them.

GammaGames commented:
Thanks for the tips kylechu! I've done some of them, you can see over at ec294bb.

Unfortunately I've got to work on classes right now, so I won't be able to work on the WeatherManager object for the next couple days at least, though I do agree it would go a long way with simplifying this.

JdotCarver commented:
Geez, so many additions! I have got to try them out! Possibly enough for the next release now.

kowsen commented:
I just remembered that I started trying to write a WeatherManager object when I first saw your code. I just fixed some bugs in it and I think got everything working - you can check it out here and take whatever ideas you want from it - e78a6d56173ddf9678ea0b93d9f45ef0493b79ed.

JdotCarver commented:
Just tell me when you're ready for the merge and I'll implement it. :)

JdotCarver commented:
Btw Please reply under what name you'd wish to appear in the credits. :)

mherndon6 commented:
Hey, I've been working on a couple features on kylechu's branch. I added some styling to the options page and added the mix all option. I'll make it compatible with GammaGames's changes and submit another pull request tonight.

GammaGames commented:
I merged twice on accident, the first automatic one added some <<<<<HEAD lines in places and I had to go through and check to make sure all the looping options were still working.
Might need to be modified a bit, I'm still using playPause(true) to star the audio

Should all be working now! I've updated adding a WeatherManager. The object checks for new weather every 10 minutes
For credit I'd like "Jesse Lieberg", thanks!

kowsen commented on scripts/StateManager.js:
```diff
@@ -67,6 +87,40 @@ function StateManager() {
+	}
+	
+	// get current weather conditions using openweathermap: http://openweathermap.org/current
+	function updateWeatherCond(zip, country, cb) {
```
updateWeatherCond is redundant and should be deleted if we have a weatherManager object

kowsen commented on scripts/StateManager.js:
```diff
@@ -25,17 +26,28 @@ function StateManager() {
 
 	// Possible events include:
-	// volume, kkStart, hourMusic, gameChange, pause
+	// volume, kkStart, hourMusic, weatherMusic, gameChange, weatherChange, pause
```
It doesn't look like there is a weatherMusic event.

kowsen commented on scripts/AudioManager.js:
```diff
@@ -83,16 +83,41 @@ function AudioManager(addEventListener, isTownTune) {
 		}
 	}
+	
+	function playPause(play) {
```
I'm confused why playPause is a separate function instead of using audio.play and audio.pause.

GammaGames replied:
That was leftovers from when I was doing things at the same time as playing or pausing, instead of using the listeners

kowsen replied:
I think it would be a good idea to remove it then so future developers aren't tempted to do something similar.

kowsen commented on scripts/AudioManager.js:
```diff
@@ -83,16 +83,41 @@ function AudioManager(addEventListener, isTownTune) {
+			audio.pause();
+		}
+	}
+	function getSrc(game, hour, weather) {
```
getSrc in AudioManager is no longer necessary since the decision is now made by StateManager and AudioManager no longer has any concept of weather.

kowsen commented on scripts/BadgeManager.js:
```diff
@@ -2,20 +2,38 @@
+		if(isEnabled()) {
+			chrome.browserAction.setBadgeText({ text: "KK" });
+		}
+		playIcon();
```
It could be cute to change the icon to KK instead of the leaf when this event comes in.

GammaGames replied:
Good idea!

kowsen commented on scripts/StateManager.js:
```diff
@@ -25,17 +26,28 @@ function StateManager() {
+		if(!weatherManager) {
+			weatherManager = new WeatherManager(options.zipCode, options.countryCode);
+			weatherManager.registerChangeCallback(function() {
+				if(!isKK() && options.music == 'new-leaf-live') {
```
You may want to change "options.music == 'new-leaf-live'" to "isLive()" so you won't have to change this conditional if we add more live weather music options.

kowsen commented on scripts/WeatherManager.js:
```diff
@@ -0,0 +1,72 @@
+	};
+
+	weatherCheckLoop();
+}
```
I would consider re-adding the debug function I had here to test weather changes. It was super helpful when I was debugging the extension. Here it is if you don't have a copy of it -

if(DEBUG_FLAG) {
   window.changeWeather = function(newWeather) {
      weather = newWeather;
      callback();
   }
}

kowsen commented on scripts/WeatherManager.js:
```diff
@@ -0,0 +1,72 @@
+ 				var response = JSON.parse(request.responseText);
+ 				if( response.cod == "200" && response.weather[0].main !== weather) {
+ 					weather = response.weather[0].main;
+ 					if( typeof callback === 'function' ) {
```
This calls the callback even if the weather changes from "Drizzle" to "Mist", even though both are just considered "Rain". You would probably want to store the result of getWeather() to a variable like "oldWeather" between lines 57 and 58, and then add a check for if getWeather() !== oldWeather in this if statement surrounding the callback.

This was my bad - should have caught it earlier.

Note: Because of how scoping is, you'll need to declare variable "self" equal to "this" at the top of the object, and then call self.getWeather() here. Look at StateManager.js lines 8 and 172 for an example of this same process.

GammaGames replied:
Good catch! I'm having a little bit of a difficulty getting the getWeather() function working. I've tried using getWeather(), this.getWeather(), window[getWeather](), none seem to work and all give a "getWeather is not defined" error

kowsen replied:
Did you follow the instructions I gave in the last paragraph of the comment above?

GammaGames replied:
Oh, the note hadn't appeared when I originally tried. Thanks for the edit, I wasn't aware of some of the strange things js does with scope when using events!

kowsen commented on scripts/BadgeManager.js:
```diff
@@ -2,20 +2,38 @@
+	
+	function playIcon() {
+		chrome.browserAction.setIcon({
+			path : "img/icon_38_leaf-02.png"
```
Is the play icon supposed to show the pause symbol and the pause icon supposed to show the play symbol? I can see that it might be saying that while music is playing, it functions as a pause button, but I feel like the other way around would be more intuitive. That's just a matter of taste though.

GammaGames replied:
The function names are confusing. I renamed them to be clearer, and made it show KK during his time.

The reason I show the pause icon while playing is because that seems to be the general standard across websites. YouTube, Spotify, and Pandora all show the opposite state as if to indicate "click here to do the thing"

GammaGames commented:
Thanks for the tips again, you've been incredibly helpful in getting this clean and nice to work with :)
As far as I can see the features are working, but you should probably wait for kylechu's go ahead first since he obviously knows his stuff!

kowsen commented on scripts/BadgeManager.js:
```diff
@@ -2,20 +2,44 @@
+		pause();
 	});
+	
+	function play(kk) {
```
I'm not the biggest fan of the functions being called "play" and "pause", since that makes them sound like they're controlling the music. Maybe have one function called changeIcon that takes two variables - isPlaying and isKK?

This is just a minor nitpick though.

kowsen commented:
You're absolutely welcome, always happy to help and grateful for the confidence you've put in me! I'm happy with this pull request - everything's looking really nice! I haven't tested it thoroughly, but I will over the next few days, and it sounds like Gamma's done some tests.

The comment on the diff I made a few minutes ago is just a super minor nitpick. If Gamma agrees and wants to change it, it would make me happy, but I think it's good as-is.

GammaGames commented:
The play/pause badge icons always bugged me, so I took some inspiration from mherdon6's branch. I've only changed a few things in the latest commit:
- I removed some extra icons that were never used (19 and 128 versions for most of them)
- Changed the icons to the more New Leaf color, with 50% opacity when paused and 100% when playing.
- Renamed the png files to make more sense

I've also included a 50% opacity icon of KK for mherdon6's alternate icon branch, because his work over there is pretty cool!
These changes should not have broken anything, I'm still running it regularly and haven't found any issues.

BordenJardine commented on scripts/StateManager.js:
```diff
@@ -67,6 +87,21 @@ function StateManager() {
+	function getMusic() {
+		if(isLive()) {
+			if(weatherManager.getWeather() == "Rain")
+				return "new-leaf-raining";
```
only works in new leaf?

GammaGames replied:
The extension doesn't have any audio files for WW/CF raining, at least that I saw. It could always be added later though

JdotCarver replied:
I'll get / upload them today.

JdotCarver commented:
Got the files in mp3 format, I'll reencode them in ogg with Audition when I come back to my main rig. (With the reduced bitrate and I'll update looping times with it too)
